### ISSUE TITLE
Incorrect behavior for keepalive connections

### ISSUE THREAD

**Issue by pracj3am** [intel/asynch_mode_nginx#10]
I have found a problem with keepalive connections. After the first request the function `ngx_http_set_keepalive` is called, where the request struct is freed (`ngx_http_free_request`), which is followed by this patched code
```c
    wev = c->write;
    wev->handler = ngx_http_empty_handler;
```
[...]
```c
#if (NGX_HTTP_SSL)
    if (c->ssl) {
        ngx_ssl_free_buffer(c);
    }
#endif

    rev->handler = ngx_http_keepalive_handler;

    if (wev->active && (ngx_event_flags & NGX_USE_LEVEL_EVENT)) {
#if (NGX_HTTP_SSL)
        if (c->asynch && ngx_del_async_conn) {
            if (c->num_async_fds) {
                ngx_del_async_conn(c, NGX_DISABLE_EVENT);
                c->num_async_fds--;
            }
        }
#endif
        if (ngx_del_event(wev, NGX_WRITE_EVENT, 0) != NGX_OK) {
            ngx_http_close_connection(c);
            return;
        }
    }
```
The function `ngx_del_async_conn` is not called, because `wev->active` is 0.

Then sometimes happen that epoll is waken up with async event `EPOLLIN | EPOLLOUT`, when a new request haven't arrived yet. So `ngx_ssl_write_async_handler` is called, which sets `c->read->handler` to `ngx_http_request_handler` instead of `ngx_http_keepalive_handler`. And then when out-of-band keepalive packet arrives, the `ngx_http_request_handler` is called, which crashes, because request structure is already freed (see segfaults reported in #6 ).

I was able to fix the problem by adding
```c
    if (c->destroyed) {
        return;
    }
```
at the beginning of functions `ngx_ssl_read_async_handler` and `ngx_ssl_write_async_handler`. But I think it should be fixed by proper calling `ngx_del_async_conn`, i.e. fix the condition, when it is called.

**Comment by daweiq** [comment]
@pracj3am Thank you for the sharing. 
@nzt4567 Does this change fix your issue?

**Comment by nzt4567** [comment]
@daweiq yes I can confirm this fixed https://github.com/intel/asynch_mode_nginx/issues/6
@pracj3am thank you

**Comment by daweiq** [comment]
@pracj3am Great work on fix the issue!
I am confused on why some async-event comes before new requests come.
Do you know to reproduce this issue? 
So I can understand it by trace the code.

**Comment by pracj3am** [comment]
I don't understand it too. Unfortunately I cannot reproduce it, it does not  happen for every keep-alive connection, just sometimes.

Here are the key lines from nginx debug log:
```
018/12/11 18:31:33 [debug] 3371#0: *526 SSL to write: 5690
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL_write: -1
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL_get_error: 9
2018/12/11 18:31:33 [debug] 3371#0: *526 event timer: 326, old: 706864481, new: 706864673
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL ASYNC WANT recieved: "ngx_ssl_write"
2018/12/11 18:31:33 [debug] 3371#0: *526 ngx_ssl_async_process_fds called
[...]
2018/12/11 18:31:33 [debug] 3371#0: epoll: fd:326 ev:0004 d:00007FF04A4A75D0, async:0
2018/12/11 18:31:33 [debug] 3371#0: fd:326, EPOLLIN 0, EPOLLOUT 4
2018/12/11 18:31:33 [debug] 3371#0: fd:326, c->read->active 1, c->write->active 1
[...]
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL to write: 5690
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL_write: 5690
2018/12/11 18:31:33 [debug] 3371#0: *526 ngx_ssl_async_process_fds called
[...]
2018/12/11 18:31:33 [debug] 3371#0: *526 http close request
[...]
2018/12/11 18:31:33 [debug] 3371#0: epoll: fd:326 ev:0005 d:00007FF04A4A75D2, async:1
2018/12/11 18:31:33 [debug] 3371#0: fd:326, EPOLLIN 1, EPOLLOUT 4
2018/12/11 18:31:33 [debug] 3371#0: fd:326, c->read->active 1, c->write->active 1
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL write async handler
2018/12/11 18:31:33 [debug] 3371#0: *526 http empty handler
```

`SSL_write` is blocked on `SSL_ERROR_WANT_ASYNC`, later event EPOLLOUT arrives and `SSL_write` succeeds, request is closed and then the async event comes.

**Comment by daweiq** [comment]
Thank you for the information! 
I will try to understand why the async event comes when error happened. 
For this issue, at least we have your patch as a fix.

**Comment by pracj3am** [comment]
If I get `SSL_ERROR_WANT_ASYNC` after unsuccessful `SSL_write`, shouldn't I wait before another try until async fd is ready for writing?

Because from the debug above I see that `SSL_write` is retried when underlying connection fd is ready for writing, not async fd. So when finally async fd is ready, `SSL_write` have already been called and request cleaned up.

I have tried to change lines in `ngx_ssl_write` to
```c
    if(c->asynch && sslerr == SSL_ERROR_WANT_ASYNC) {
        c->async->ready = 0;
        c->async->handler = ngx_ssl_write_async_handler;
        if (c->write->handler != ngx_ssl_empty_handler) {
            c->write->saved_handler = c->write->handler;
            c->write->handler = ngx_ssl_empty_handler;
        }
```
and in `ngx_ssl_write_async_handler` to
```c
    aev->ready = 0;
    aev->handler = ngx_ssl_empty_handler;
    if (c->write->saved_handler != ngx_ssl_empty_handler) {
        c->write->handler = c->write->saved_handler;
        c->write->saved_handler = ngx_ssl_empty_handler;
    }

    c->write->handler(c->write);
```
and it seems that it fixes the issue too.

**Comment by daweiq** [comment]
From the log, it shows the async event timeout (fd 326) happened before async event return.  In timeout handler, it closes the http request. Later on, when async event actually returns, the async handler tried to revisit the request which has been free by timeout handler, so segment fault happens. 
This issue affect the keep-alive connection only, because in non keep-alive connection all connect related fd will be deleted when delete the connection. 
Your fix of check the c->destroy flag in async handler can fix this issue. 

One thing strange is the default NGX_ASYNC_EVENT_TIMEOUT is 10000. While, in your log the timeout happened within 1s, do you change it to <= 1s?

**Comment by pracj3am** [comment]
Hello!

`NGX_ASYNC_EVENT_TIMEOUT` was 10000 and no async event timeout happened in this case. It is explained in my previous comment. I will try to explain it in more details.

Here is the nginx debug log lines, this time I included the lines that are related to nginx timers and added my comments.
```
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL to write: 5690
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL_write: -1
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL_get_error: 9
2018/12/11 18:31:33 [debug] 3371#0: *526 event timer: 326, old: 706864481, new: 706864673
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL ASYNC WANT recieved: "ngx_ssl_write"
2018/12/11 18:31:33 [debug] 3371#0: *526 ngx_ssl_async_process_fds called
```
 `SSL_write` fails with error WANT ASYNC, so we have to wait until async file descriptor will be ready for writing and then we could call `SSL_write` again. 10s async event timer is set.
```
[...]
2018/12/11 18:31:33 [debug] 3371#0: *526 event timer add: 326: 90000:706944673
[...]
```
set event timer for 90s `send_timeout`
```
2018/12/11 18:31:33 [debug] 3371#0: epoll: fd:326 ev:0004 d:00007FF04A4A75D0, async:0
2018/12/11 18:31:33 [debug] 3371#0: fd:326, EPOLLIN 0, EPOLLOUT 4
2018/12/11 18:31:33 [debug] 3371#0: fd:326, c->read->active 1, c->write->active 1
2018/12/11 18:31:33 [debug] 3371#0: *526 http run request
[...]
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL to write: 5690
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL_write: 5690
```
An epoll event, *tcp socket* with fd:326 is ready for writing, so `c->write->handler` is called, which calls `SSL_write`, which succeeds this time. This is imo bug, calling `SSL_write` should be postponed, until also the async fd is ready for writing (and all data encrypted).
```
2018/12/11 18:31:33 [debug] 3371#0: *526 ngx_ssl_async_process_fds called
2018/12/11 18:31:33 [debug] 3371#0: *526 http write filter 0000000000000000
2018/12/11 18:31:33 [debug] 3371#0: *526 http copy filter: 0 
2018/12/11 18:31:33 [debug] 3371#0: *526 http writer output filter: 0, 
2018/12/11 18:31:33 [debug] 3371#0: *526 http writer done: 
2018/12/11 18:31:33 [debug] 3371#0: *526 http finalize request: 0, 
2018/12/11 18:31:33 [debug] 3371#0: *526 event timer del: 326: 706944673
2018/12/11 18:31:33 [debug] 3371#0: *526 set http keepalive handler
2018/12/11 18:31:33 [debug] 3371#0: *526 http close request
[...]
```
All request data have been written, so the http request is closed and related structured are going to be freed. The `send_timeout` event timer is deleted, `c->read->handler` is set to `ngx_http_keepalive_handler`. 
```
2018/12/11 18:31:33 [debug] 3371#0: epoll: fd:326 ev:0005 d:00007FF04A4A75D2, async:1
2018/12/11 18:31:33 [debug] 3371#0: fd:326, EPOLLIN 1, EPOLLOUT 4
2018/12/11 18:31:33 [debug] 3371#0: fd:326, c->read->active 1, c->write->active 1
2018/12/11 18:31:33 [debug] 3371#0: *526 SSL write async handler
2018/12/11 18:31:33 [debug] 3371#0: *526 http empty handler
```
In the meantime Intel® card also finished the encryption (async fd is ready for writing), `c->async->handler` is called, `ngx_ssl_write_async_handler` sets `c->read->handler` to `ngx_http_request_handler` instead of `ngx_http_keepalive_handler`. And then when out-of-band keepalive packet arrives, the `ngx_http_request_handler` is called -> segfault.

**Comment by pracj3am** [comment]
`c->write->handler` is set to `ngx_http_request_handler`, which calls `r->write_event_handler`, which is set to `ngx_http_writer`, which calls `ngx_http_output_filter` [...] which calls `ngx_ssl_send_chain`.

Why does `SSL_write` succeed then? I suppose that before nginx actually calls the `SSL_write` (it processes output buffers through all output filters), in the meantime the Intel card finishes an async encrypt job.

**Comment by daweiq** [comment]
Thank you for the information!
We will include your proposal in the master to fix it. 
```
    if (c->destroyed) {
        return;
    }
```
